src/pip/_internal/utils/pkg_resources.py

```diff
@@ -0,0 +1,44 @@
+from pip._vendor.pkg_resources import yield_lines
+from pip._vendor.six import ensure_str
+
+from pip._internal.utils.typing import MYPY_CHECK_RUNNING
+
+if MYPY_CHECK_RUNNING:
+    from typing import Dict, Iterable, List
+
+
+class DictMetadata(object):
+    """IMetadataProvider that reads metadata files from a dictionary.
+    """
+    def __init__(self, metadata):
+        # type: (Dict[str, bytes]) -> None
+        self._metadata = metadata
+
+    def has_metadata(self, name):
+        # type: (str) -> bool
+        return name in self._metadata
+
+    def get_metadata(self, name):
+        # type: (str) -> str
+        try:
+            return ensure_str(self._metadata[name])
```

Comment:
I was thinking if there's much point in doing this on metadata-access vs metadata-load.

Is there any specific reason you've done it this way? I feel it might be cleaner to do "ensure_str" on everything on construction -- a successfully constructed `DictMetadata` would never (intentionally) raise errors via any method then.

Reply:
I originally had done it as you describe, but @sbidoul pointed out [here](https://github.com/pypa/pip/pull/7539#discussion_r362964913) that this breaks at least one package. The current construction with delayed decoding should have exactly the same behavior, exception-wise, as the original code.

An alternative non-breaking approach would be to eagerly decode a set of known files (`METADATA`, `WHEEL`, `entry_points.txt`, `RECORD`). Since this will have implications on both the transition to importlib.metadata and the resolver work (a parse failure leading us to choose another wheel, for example), I think it's best to defer the behavior change until we implement those.

Reply:
Sure, that sounds reasonable to me. :)